steering_vectors/utils.py

```diff
@@ -21,3 +25,38 @@ def batchify(
         desc=tqdm_desc,
     ):
         yield data[i : i + batch_size]
+
+
+def get_default_device() -> str:
+    if torch.cuda.is_available():
+        return "cuda"
+    elif torch.backends.mps.is_available() and torch.backends.mps.is_built():
+        # Parse the PyTorch version to check if it's below version 2.0
+        major_version = int(torch.__version__.split(".")[0])
+        if major_version >= 2:
+            return "mps"
+    else:
+        return "cpu"
+
+    raise RuntimeError("Should not reach here!")
+
+
+@Singleton
+class DeviceManager:
```

Comment:
It doesn't looks like this is ever being used. Why is this needed?

Reply:
I'm using this in the examples

Reply:
Why not just set the device explicitly in examples, e.g. put `DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'` at the top of the notebook? It feels odd to put example code into the public API of a library IMO

Reply:
fair. I agree with that, but I also think this is just a useful pattern generally to control instantiation of new tensors. The nice thing about this is that you can nest multiple contexts, within which different defaults will be used. If it turns out we really don't need this outside the examples I'll remove it

Reply:
I can see how it would be used, but IMO if this is just for examples it's more confusing than just writing out the code to set the device in the notebook. This is a non-standard object we've created, and will require users to read the actual code to understand what it's doing. Personally, I'd think it's cleaner to just define this object in the notebook in that case, so then users don't need to look at our source code to understand what `DeviceManager.instance()` means, no? Plus it's relying on decorators and the singleton pattern, which is more things users need to understand rather than a simple `x.device = 'cuda'` which should be pretty clear